opsdroid/core.py

```diff
@@ -97,54 +107,95 @@ def critical(self, error, code):
         _LOGGER.critical(error)
         self.exit()
 
-    def call_stop(self):
-        """Signal handler to call disconnect and stop."""
-        future = asyncio.ensure_future(self.disconnect())
-        future.add_done_callback(self.stop)
-        return future
+    @staticmethod
+    def handle_async_exception(loop, context):
+        """Handle exceptions from async coroutines."""
+        _LOGGER.error("Caught exception")
+        _LOGGER.error(context)
+
+    def is_running(self):
+        """Check whether opsdroid is running."""
+        return self._running
+
+    async def handle_signal(self):
+        """Handle signals."""
+        self._running = False
+        await self.unload()
+
+    def run(self):
+        """Start the event loop."""
+        _LOGGER.info(_("Opsdroid is now running, press ctrl+c to exit."))
+        if not self.is_running():
+            self._running = True
+            while self.is_running():
+                pending = asyncio.Task.all_tasks()
+                try:
+                    self.eventloop.run_until_complete(asyncio.gather(*pending))
+                except asyncio.CancelledError:
+                    pass
+
+            self.eventloop.stop()
+            self.eventloop.close()
+
+            _LOGGER.info(_("Bye!"))
+            self.exit()
+        else:
+            _LOGGER.error(_("Oops! Opsdroid is already running."))
+
+    def load(self):
+        """Load modules."""
+        self.modules = self.loader.load_modules_from_config(self.config)
+        _LOGGER.debug(_("Loaded %i skills"), len(self.modules["skills"]))
+        self.setup_skills(self.modules["skills"])
+        self.web_server = Web(self)
+        self.web_server.setup_webhooks(self.skills)
+        self.train_parsers(self.modules["skills"])
+        if self.modules["databases"] is not None:
+            self.start_databases(self.modules["databases"])
+        self.start_connectors(self.modules["connectors"])
+        self.cron_task = self.eventloop.create_task(parse_crontab(self))
+        self.eventloop.create_task(self.web_server.start())
+
+    async def unload(self, future=None):
+        """Stop the event loop."""
+        _LOGGER.info(_("Received stop signal, exiting."))
+
+        _LOGGER.info(_("Removing skills..."))
+        for skill in self.skills:
+            _LOGGER.info(_("Removed %s"), skill.config['name'])
+            self.skills.remove(skill)
 
-    async def disconnect(self):
-        """Disconnect all the connectors."""
         for connector in self.connectors:
+            _LOGGER.info(_("Stopping connector %s..."), connector.name)
             await connector.disconnect(self)
+            self.connectors.remove(connector)
```

Comment:
_Note: This is a thought._
Perhaps we could somehow change the infinite loop flag from `connector.listen` to false to avoid the exception `Task pending coro=<ConnectorShell.listen() running at (...)`? Could perhaps a simple connector.is_listening = False work?

Reply:
The shell connector just needs to implement the disconnect method correctly.